File: P3_Mojo/data/Naive_Bayes/Python/Naive_Bayes.py
import numpy as np
import csv
import time

class NaiveBayes:
    def __init__(self):
        self.classes = None
        self.class_priors = None
        self.feature_probs = None
    
    def fit(self, X, y):
        num_samples, num_features = X.shape
        self.classes = np.unique(y)
        num_classes = len(self.classes)
        
        self.class_priors = np.zeros(num_classes)
        self.feature_probs = np.zeros((num_classes, num_features))
        
        for i, c in enumerate(self.classes):
            X_c = X[y == c]
            self.class_priors[i] = len(X_c) / num_samples
            self.feature_probs[i] = np.mean(X_c, axis=0)
    
    def predict(self, X):
        num_samples, num_features = X.shape
        num_classes = len(self.classes)
        
        predictions = []
        for i in range(num_samples):
            sample_probs = np.zeros(num_classes)
            for j in range(num_classes):
                class_prior = self.class_priors[j]
                feature_probs = self.feature_probs[j]
                sample_probs[j] = np.sum(np.log(feature_probs + 1e-9) * X[i] + np.log(1 - feature_probs + 1e-9) * (1 - X[i])) + np.log(class_prior)
            predicted_class = self.classes[np.argmax(sample_probs)]
            predictions.append(predicted_class)
        
        return predictions

start_time = time.time()

# Leer datos de entrenamiento desde archivo CSV
X_train = []
y_train = []
with open('assets/naive_bayes.csv', 'r') as archivo:
    lector_csv = csv.reader(archivo)
    for fila in lector_csv:
        X_train.append(fila[:-1])  # Todos los valores de la fila excepto el último
        y_train.append(fila[-1])   # Último valor de la fila

# Convertir a arrays de NumPy
X_train = np.array(X_train, dtype=int)
y_train = np.array(y_train)

# Crear y ajustar el modelo Naive Bayes
model = NaiveBayes()
model.fit(X_train, y_train)

# Datos de prueba
X_test = np.array([[1,1,1,1,1,1,1,1,1,0,0,1,1,0,0,1,1,0,0,0,0,0,1,1,0,0,0,1,1,0,0,1,0,0,0,1,0,1,0,0,1,0,0,1,0,0,1,0,0,1,0,0,0,0,1,0,1,0,1,1,0,0,0,0,0,1,1,0,0,1,0,1,0,0,1,0,1,1,1,0,0,1,1,0,1,0,1,0,1,1,0,0,1,1,1,0,1,1,1,1],
                   [1,1,0,1,0,1,1,1,1,0,1,0,0,0,1,0,0,0,0,1,1,0,1,1,1,0,0,1,0,1,0,0,1,0,1,1,1,1,0,0,0,0,1,0,1,0,1,1,0,0,1,1,0,0,0,0,0,1,1,0,0,1,1,0,1,0,1,0,1,1,1,1,0,1,1,0,1,1,0,1,0,0,0,1,1,1,1,0,0,0,1,1,1,0,0,1,0,0,1,0]])

# Realizar predicciones
y_pred = model.predict(X_test)

# Fin del procesamiento
end_time = time.time()

# Calcular tiempo de procesamiento
processing_duration = end_time - start_time

print("Tiempo de procesamiento total: {:.2f} segundos".format(processing_duration))

print("Predicciones:", y_pred)

File: P3_Mojo/data/Naive_Bayes/CPP/Naive_Bayes.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <cmath>
#include <sstream>
#include <algorithm>

class NaiveBayes {
public:
    void fit(const std::vector<std::vector<int>>& X, const std::vector<std::string>& y) {
        num_samples_ = X.size();
        num_features_ = X[0].size();

        classes_ = getUniqueClasses(y);
        num_classes_ = classes_.size();

        class_priors_.resize(num_classes_);
        feature_probs_.resize(num_classes_, std::vector<double>(num_features_));

        for (int i = 0; i < num_classes_; i++) {
            const std::string& c = classes_[i];
            std::vector<std::vector<int>> X_c;
            for (int j = 0; j < num_samples_; j++) {
                if (y[j] == c) {
                    X_c.push_back(X[j]);
                }
            }

            class_priors_[i] = static_cast<double>(X_c.size()) / num_samples_;

            for (int f = 0; f < num_features_; f++) {
                int sum_feature = 0;
                for (int j = 0; j < X_c.size(); j++) {
                    sum_feature += X_c[j][f];
                }
                feature_probs_[i][f] = static_cast<double>(sum_feature) / X_c.size();
            }
        }
    }

    std::vector<std::string> predict(const std::vector<std::vector<int>>& X) {
        std::vector<std::string> predictions;
        for (const auto& sample : X) {
            std::vector<double> sample_probs(num_classes_);
            for (int i = 0; i < num_classes_; i++) {
                double class_prior = class_priors_[i];
                const std::vector<double>& feature_probs = feature_probs_[i];
                double log_prob = 0.0;
                for (int f = 0; f < num_features_; f++) {
                    if (sample[f] == 1) {
                        log_prob += std::log(feature_probs[f]);
                    } else {
                        log_prob += std::log(1.0 - feature_probs[f]);
                    }
                }
                sample_probs[i] = log_prob + std::log(class_prior);
            }
            int max_index = std::max_element(sample_probs.begin(), sample_probs.end()) - sample_probs.begin();
            predictions.push_back(classes_[max_index]);
        }
        return predictions;
    }

private:
    std::vector<std::string> getUniqueClasses(const std::vector<std::string>& y) {
        std::vector<std::string> unique_classes;
        for (const auto& cls : y) {
            if (std::find(unique_classes.begin(), unique_classes.end(), cls) == unique_classes.end()) {
                unique_classes.push_back(cls);
            }
        }
        return unique_classes;
    }

    int num_samples_;
    int num_features_;
    int num_classes_;
    std::vector<std::string> classes_;
    std::vector<double> class_priors_;
    std::vector<std::vector<double>> feature_probs_;
};


// Función para verificar si una cadena es numérica
bool isNumeric(const std::string& str) {
    for (char c : str) {
        if (!std::isdigit(c)) {
            return false;
        }
    }
    return true;
}

// Función para leer los datos de las columnas excepto la última
std::vector<std::vector<int>> readCSV(const std::string& filename) {
    std::ifstream file(filename);
    std::vector<std::vector<int>> data;
    std::string line;

    while (std::getline(file, line)) {
        std::vector<int> row;
        std::istringstream iss(line);
        std::string value;

        while (std::getline(iss, value, ',')) {
            // Verificar si el valor es numérico antes de convertirlo a entero
            if (isNumeric(value)) {
                row.push_back(std::stoi(value));
            }
        }

        // Ignorar la última columna
        row.pop_back();

        data.push_back(row);
    }

    file.close();

    return data;
}

// Función para leer la última columna (etiquetas)
std::vector<std::string> readLabelsCSV(const std::string& filename) {
    std::ifstream file(filename);
    std::vector<std::string> labels;
    std::string line;

    while (std::getline(file, line)) {
        std::istringstream iss(line);
        std::string value;

        while (std::getline(iss, value, ',')) {
            // Leer únicamente la última columna
            labels.push_back(value);
        }
    }

    file.close();

    return labels;
}
void savePredictions(const std::vector<std::string>& predictions, const std::string& filename) {
    std::ofstream file(filename);
    for (const auto& prediction : predictions) {
        file << prediction << "\n";
    }
    file.close();
}

int main() {
    // Leer datos de entrenamiento desde archivo CSV
    std::vector<std::vector<int>> X_train = readCSV("../../assets/naive_bayes.csv");
    std::vector<std::string> y_train = readLabelsCSV("../../assets/naive_bayes.csv");

    // Crear y ajustar el modelo Naive Bayes
    NaiveBayes model;
    model.fit(X_train, y_train);

    // Datos de prueba
    std::vector<std::vector<int>> X_test = {
        {1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1, 1, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 1, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0, 1, 0, 1, 1, 1, 0, 0, 1, 1, 0, 1, 0, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 1, 1, 1, 1},
        {1, 1, 0, 1, 0, 1, 1, 1, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0, 0, 1, 0, 1, 1, 1, 1, 0, 0, 0, 0, 1, 0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 1, 0, 1, 0, 1, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1, 1, 0, 1, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 1, 0}
    };

    // Realizar predicciones
    std::vector<std::string> y_pred = model.predict(X_test);

    // Guardar las predicciones en un archivo
    savePredictions(y_pred, "predictions.txt");

    std::cout << "Predicciones guardadas en predictions.txt" << std::endl;

    return 0;
}
